# **Understanding the Business Context**

## **Business Context**
Greenhouse gas (GHG) emissions have a profound impact on climate change, affecting global temperatures, sea levels, and weather patterns. Governments, policymakers, and businesses rely on emissions data to formulate policies, set sustainability goals, and monitor progress toward climate commitments.

## **Business Need**
There is a growing demand for data-driven insights into emissions trends, sector-wise contributions, and the effectiveness of sustainability policies. This project will help answer:

✔ How have global emissions evolved over time?

✔ Which sectors contribute the most?

✔ How can we forecast future emissions trends?

✔ What actionable insights can be derived to reduce emissions?

✔ Which countries or regions have contributed the most to emissions over time?

✔ How effective have policy changes been in reducing emissions?

## **Business Process Behind the Request**
This type of data is typically used by:

Policymakers → For setting climate goals and regulations

Corporations → For sustainability reporting and reducing carbon footprints

Researchers & NGOs → For analyzing environmental impact and tracking global trends

# 🌍 Greenhouse Gas Emissions Analysis & Modeling

## **📌 Step 1: Business Understanding & Defining Goals**  
### **1.1 Business Problem**  
We aim to analyze historical greenhouse gas (GHG) emissions to:  
- **Track emission trends** over time.  
- **Evaluate policy impact** on emissions.  
- **Predict future emission levels** using machine learning.  
- **Support decision-making** for sustainability initiatives.  

### **1.2 Intended Audience**  
👥 The following groups will use this analysis:  
- **Government & Policymakers** – for emission regulations.  
- **Businesses & Corporations** – for ESG compliance.  
- **Sustainability Analysts & NGOs** – to assess policy effectiveness.  
- **Researchers & Academics** – for climate change studies.  

### **1.3 Deliverables**  
📌 Our final outputs will include:  
✅ **Power BI, Tableau, StoryMaps Dashboards**  
✅ **Summary Reports** (Google Docs, Notion)  
✅ **Predictive Models** (Python, Machine Learning)  
✅ **Infographics & Presentations**  

### **1.4 Key Performance Indicators (KPIs)**  
📊 We will measure success using:  
- **Total CO2 emissions per country/sector/year**  
- **Percentage change in emissions over time**  
- **Per capita emissions** (CO2 per person)  
- **Emissions intensity** (CO2 per unit of GDP)  
- **Forecast accuracy for emissions prediction models**  

### **1.5 Business Processes Behind the Request**  
🔎 The need for this analysis comes from:  
- **Paris Agreement & Net-Zero Goals 2050**  
- **Government GHG accounting regulations**  
- **Business carbon footprint tracking for ESG reporting**  

# **Understanding the Data**

## **Data Source**

Climate Watch (Historical GHG Emissions Data, 1990-2021)

## **📊 Nature of the Data**

Structured dataset containing greenhouse gas emissions across different years, regions, and sectors

Includes CO₂, CH₄, N₂O, and other GHG emissions data

Covers various economic sectors (Energy, Transport, Industry, Agriculture, etc.)


In [1]:
import pandas as pd  

# Load Data
file_path = "GHG_Emissions.csv"  
df = pd.read_csv(file_path)  

# Display first 5 rows
df.head()

ISO Country    Data source                Sector      Gas    Unit  \
0  WORLD   World  Climate Watch  Total including LUCF  All GHG  MtCO₂e   
1  WORLD   World  Climate Watch  Total excluding LUCF  All GHG  MtCO₂e   
2  WORLD   World  Climate Watch                Energy  All GHG  MtCO₂e   
3  WORLD   World  Climate Watch  Total including LUCF      CO2  MtCO₂e   
4  WORLD   World  Climate Watch  Total excluding LUCF      CO2  MtCO₂e   

       2021      2020      2019      2018  ...      1999      1998      1997  \
0  49553.48  47463.17  49843.57  49482.10  ...  35281.42  35278.92  35739.15   
1  48209.50  46066.32  48046.60  47960.07  ...  33346.20  33152.61  32942.62   
2  37407.79  35450.68  37613.05  37635.61  ...  25466.40  25336.58  25177.64   
3  36693.41  34819.59  37046.27  36865.29  ...  25227.59  25154.68  25546.76   
4  35540.43  33646.42  35575.61  35569.42  ...  23553.91  23384.67  23220.76   

       1996      1995      1994      1993      1992      1991      1990  
0  34372.44  33982.29  33191.77  32924.03  32787.52  32880.80  32735.02  
1  32545.41  31949.87  31164.96  30896.88  30760.80  30854.08  30708.31  
2  24756.95  24245.42  23592.68  23503.34  23380.55  23505.38  23364.85  
3  24434.73  24086.41  23457.62  23311.16  23167.98  23217.45  23060.38  
4  22798.57  22317.43  21694.25  21547.45  21404.71  21454.18  21297.11  

[5 rows x 38 columns]

## **Features in the Data:**

1. Country/Region – Specifies whether the data represents the world or a specific country
2. Data Source – "World Climate Watch"
3. Sector – The economic or industrial sector contributing to emissions, such as:
* Energy – Emissions from energy production and consumption.
* Electricity/Heat – Emissions from power generation.
* Total (Including/Excluding LUCF) – Overall emissions, with or without Land Use Change and Forestry (LUCF) contributions.
4. Gas Type – The type of greenhouse gas being measured, such as:
* GHG (Greenhouse Gases)
* CO₂ (Carbon Dioxide)
* Methane (CH₄)
* Nitrous Oxide (N₂O)
5. Unit of Measurement – in Million Metric Tons of CO₂ equivalent (MtCO₂e).
6. Yearly Emission Data – Numbers indicating emissions levels across different years from 1990 to 2021.

In [2]:
# Check dataset structure
print("\nDataset Info:")
print(df.info())


Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10928 entries, 0 to 10927
Data columns (total 38 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   ISO          10928 non-null  object 
 1   Country      10928 non-null  object 
 2   Data source  10928 non-null  object 
 3   Sector       10928 non-null  object 
 4   Gas          10928 non-null  object 
 5   Unit         10928 non-null  object 
 6   2021         10923 non-null  float64
 7   2020         10924 non-null  float64
 8   2019         10923 non-null  float64
 9   2018         10924 non-null  float64
 10  2017         10924 non-null  float64
 11  2016         10926 non-null  float64
 12  2015         10924 non-null  float64
 13  2014         10924 non-null  float64
 14  2013         10924 non-null  float64
 15  2012         10918 non-null  float64
 16  2011         10893 non-null  float64
 17  2010         10891 non-null  float64
 18  2009         10889 non-null  fl

In [3]:
# Summary statistics
print("\nSummary Statistics:")
print(df.describe())


Summary Statistics:
               2021          2020          2019          2018          2017  \
count  10923.000000  10924.000000  10923.000000  10924.000000  10924.000000   
mean      69.249856     66.097739     69.420656     69.219737     67.762867   
std     1056.653203   1004.748301   1054.656929   1051.496841   1025.740185   
min     -647.850000   -647.850000   -647.850000   -647.910000   -647.910000   
25%        0.010000      0.010000      0.010000      0.010000      0.010000   
50%        0.420000      0.400000      0.450000      0.425000      0.420000   
75%        6.230000      6.110000      6.250000      6.172500      6.012500   
max    49553.480000  47463.170000  49843.570000  49482.100000  48341.290000   

               2016          2015          2014          2013          2012  \
count  10926.000000  10924.000000  10924.000000  10924.000000  10918.000000   
mean      66.792576     66.123134     66.241657     65.750736     64.882562   
std     1010.350805   1003.799

In [4]:
# Check for missing values
print("\n Missing Values Per Column:")
print(df.isnull().sum())


 Missing Values Per Column:
ISO              0
Country          0
Data source      0
Sector           0
Gas              0
Unit             0
2021             5
2020             4
2019             5
2018             4
2017             4
2016             2
2015             4
2014             4
2013             4
2012            10
2011            35
2010            37
2009            39
2008            41
2007            40
2006            40
2005            40
2004            65
2003            65
2002            65
2001            67
2000            67
1999           139
1998           139
1997           139
1996           140
1995           142
1994           165
1993           167
1992           179
1991           231
1990           284
dtype: int64


Observations from Data Exploration
1.	Dataset Size: 10,928 rows × 38 columns.
2.	Missing Values:
* Some years (especially 2010-2021) have a few missing values.
3.	Outliers:
* Some negative emissions values (e.g., -647.85, -707.98) could be errors.
4.	Data Structure:
* Each row contains ISO, Country, Sector, Gas, and yearly emissions data.
5.	Unit Standardization Needed:
* Need to check if all values are in metric tons of CO2-equivalent (MtCO₂e).

## ** Handling Missing Values**

We will fill missing numeric values using forward fill (ffill()) since emissions usually follow a trend over time.

In [5]:
# Forward-fill missing values
df.fillna(method='ffill', inplace=True)

# Check again for missing values
print("\nMissing Values After Cleaning:")
print(df.isnull().sum())


Missing Values After Cleaning:
ISO            0
Country        0
Data source    0
Sector         0
Gas            0
Unit           0
2021           0
2020           0
2019           0
2018           0
2017           0
2016           0
2015           0
2014           0
2013           0
2012           0
2011           0
2010           0
2009           0
2008           0
2007           0
2006           0
2005           0
2004           0
2003           0
2002           0
2001           0
2000           0
1999           0
1998           0
1997           0
1996           0
1995           0
1994           0
1993           0
1992           0
1991           0
1990           0
dtype: int64


/var/folders/rg/j1hzfkkx1fn412ygjvnwy86h0000gn/T/ipykernel_77281/2667180110.py:2: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.fillna(method='ffill', inplace=True)


## **Remove Negative & Unrealistic Values**
* Negative emissions could be data entry errors or carbon capture projects.
* We will replace extreme negatives with NaN and forward-fill again.

In [6]:
# Replace extreme negatives with NaN
df[df.iloc[:, 6:] < 0] = None  # All year columns (2021-1990)

# Forward-fill again
df.fillna(method='ffill', inplace=True)

# Check if negatives still exist
print("\n📌 Min Values in Each Year Column After Cleaning:")
print(df.iloc[:, 6:].min())


📌 Min Values in Each Year Column After Cleaning:
2021    0.0
2020    0.0
2019    0.0
2018    0.0
2017    0.0
2016    0.0
2015    0.0
2014    0.0
2013    0.0
2012    0.0
2011    0.0
2010    0.0
2009    0.0
2008    0.0
2007    0.0
2006    0.0
2005    0.0
2004    0.0
2003    0.0
2002    0.0
2001    0.0
2000    0.0
1999    0.0
1998    0.0
1997    0.0
1996    0.0
1995    0.0
1994    0.0
1993    0.0
1992    0.0
1991    0.0
1990    0.0
dtype: float64


/var/folders/rg/j1hzfkkx1fn412ygjvnwy86h0000gn/T/ipykernel_77281/4148864420.py:5: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.fillna(method='ffill', inplace=True)


## Standardize Column Names & Data Types
* Rename columns to lowercase and remove spaces.
* Ensure numeric data types are correct.

In [7]:
# Rename columns
#df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')

# Convert year columns to float
for col in df.columns[6:]:
    df[col] = pd.to_numeric(df[col], errors='coerce')

print("\n Data Types After Cleaning:")
print(df.dtypes)


 Data Types After Cleaning:
ISO             object
Country         object
Data source     object
Sector          object
Gas             object
Unit            object
2021           float64
2020           float64
2019           float64
2018           float64
2017           float64
2016           float64
2015           float64
2014           float64
2013           float64
2012           float64
2011           float64
2010           float64
2009           float64
2008           float64
2007           float64
2006           float64
2005           float64
2004           float64
2003           float64
2002           float64
2001           float64
2000           float64
1999           float64
1998           float64
1997           float64
1996           float64
1995           float64
1994           float64
1993           float64
1992           float64
1991           float64
1990           float64
dtype: object


## Remove Duplicate Rows

In [8]:
# Check for duplicates
print("\n Duplicate Rows Before Cleaning:", df.duplicated().sum())

# Drop duplicates
df.drop_duplicates(inplace=True)

print("\n Duplicate Rows After Cleaning:", df.duplicated().sum())


 Duplicate Rows Before Cleaning: 0

 Duplicate Rows After Cleaning: 0


## Save the Cleaned Data

In [9]:
# Save cleaned dataset
# Identify fixed columns (non-year columns)
fixed_columns = ['ISO', 'Country', 'Data source', 'Sector', 'Gas', 'Unit']

# Convert wide format (years as columns) → long format
df_long = df.melt(id_vars=fixed_columns, var_name='Year', value_name='Emissions')

# Convert Year column to integer
df_long['Year'] = pd.to_numeric(df_long['Year'], errors='coerce')

# Keep only relevant years (1990-2021)
df_long = df_long[(df_long['Year'] >= 1990) & (df_long['Year'] <= 2021)]

# Save the reshaped dataset
print("\nGHG Emissions dataset has been cleaned and reshaped to long format and saved as 'cleaned_emissions.csv'.")

df_long.to_csv("cleaned_emissions.csv", index=False)


GHG Emissions dataset has been cleaned and reshaped to long format and saved as 'cleaned_emissions.csv'.


## **Population and GDP Datasets**

In [11]:
# Load Population Data
pop_file = "Population.csv"  
gdp_file = "GDP.csv" 

In [12]:
population = pd.read_csv(pop_file)
gdp = pd.read_csv(gdp_file)

In [13]:
population.head()

Country Name Country Code     Indicator Name Indicator Code  \
0                        Aruba          ABW  Population, total    SP.POP.TOTL   
1  Africa Eastern and Southern          AFE  Population, total    SP.POP.TOTL   
2                  Afghanistan          AFG  Population, total    SP.POP.TOTL   
3   Africa Western and Central          AFW  Population, total    SP.POP.TOTL   
4                       Angola          AGO  Population, total    SP.POP.TOTL   

          1960         1961         1962         1963         1964  \
0      54922.0      55578.0      56320.0      57002.0      57619.0   
1  130072080.0  133534923.0  137171659.0  140945536.0  144904094.0   
2    9035043.0    9214083.0    9404406.0    9604487.0    9814318.0   
3   97630925.0   99706674.0  101854756.0  104089175.0  106388440.0   
4    5231654.0    5301583.0    5354310.0    5408320.0    5464187.0   

          1965  ...         2015         2016         2017         2018  \
0      58190.0  ...     107906.0     108727.0     108735.0     108908.0   
1  149033472.0  ...  607123269.0  623369401.0  640058741.0  657801085.0   
2   10036008.0  ...   33831764.0   34700612.0   35688935.0   36743039.0   
3  108772632.0  ...  418127845.0  429454743.0  440882906.0  452195915.0   
4    5521981.0  ...   28157798.0   29183070.0   30234839.0   31297155.0   

          2019         2020         2021         2022         2023  \
0     109203.0     108587.0     107700.0     107310.0     107359.0   
1  675950189.0  694446100.0  713090928.0  731821393.0  750503764.0   
2   37856121.0   39068979.0   40000412.0   40578842.0   41454761.0   
3  463365429.0  474569351.0  485920997.0  497387180.0  509398589.0   
4   32375632.0   33451132.0   34532429.0   35635029.0   36749906.0   

   Unnamed: 68  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  

[5 rows x 69 columns]

In [14]:
gdp.head()

Country Name Country Code     Indicator Name  \
0                        Aruba          ABW  GDP (current US$)   
1  Africa Eastern and Southern          AFE  GDP (current US$)   
2                  Afghanistan          AFG  GDP (current US$)   
3   Africa Western and Central          AFW  GDP (current US$)   
4                       Angola          AGO  GDP (current US$)   

   Indicator Code          1960          1961          1962          1963  \
0  NY.GDP.MKTP.CD           NaN           NaN           NaN           NaN   
1  NY.GDP.MKTP.CD  2.421063e+10  2.496398e+10  2.707880e+10  3.177575e+10   
2  NY.GDP.MKTP.CD           NaN           NaN           NaN           NaN   
3  NY.GDP.MKTP.CD  1.190495e+10  1.270788e+10  1.363076e+10  1.446909e+10   
4  NY.GDP.MKTP.CD           NaN           NaN           NaN           NaN   

           1964          1965  ...          2015          2016          2017  \
0           NaN           NaN  ...  2.962907e+09  2.983635e+09  3.092429e+09   
1  3.028579e+10  3.381317e+10  ...  8.982778e+11  8.289428e+11  9.729989e+11   
2           NaN           NaN  ...  1.913422e+10  1.811657e+10  1.875346e+10   
3  1.580376e+10  1.692109e+10  ...  7.717669e+11  6.943610e+11  6.878492e+11   
4           NaN           NaN  ...  9.049642e+10  5.276162e+10  7.369015e+10   

           2018          2019          2020          2021          2022  \
0  3.276184e+09  3.395799e+09  2.481857e+09  2.929447e+09  3.279344e+09   
1  1.012306e+12  1.009721e+12  9.333918e+11  1.085745e+12  1.191423e+12   
2  1.805322e+10  1.879944e+10  1.995593e+10  1.426000e+10  1.449724e+10   
3  7.704950e+11  8.264838e+11  7.898017e+11  8.493124e+11  8.839739e+11   
4  7.945069e+10  7.089796e+10  4.850156e+10  6.650513e+10  1.043997e+11   

           2023  Unnamed: 68  
0  3.648573e+09          NaN  
1  1.245472e+12          NaN  
2  1.723305e+10          NaN  
3  7.991060e+11          NaN  
4  8.482465e+10          NaN  

[5 rows x 69 columns]

Key Points from the Data:
1.	Columns:
* Country Name: Full country name
* Country Code: ISO-3 country code
* Indicator Name: Specifies what the data represents (e.g., “Population, total”, “GDP (current US$)”)
* Indicator Code: Unique identifier for each type of data (e.g., SP.POP.TOTL for population, NY.GDP.MKTP.CD for GDP)
* Years (1960-2023): Each column represents the population or GDP value for that year.
2.	Missing Data:
* Some countries have missing values (NaN) for certain years, especially for GDP.
* Unnamed: 68 (extra column) is an artifact from the CSV format and can be removed.
3.	We only need data from 1990 to 2021, so we will filter out unnecessary years (1960-1989 & 2022-2023).


Before filtering and reshaping the data, we should perform a data quality check to identify and handle:

✅ Null or missing values

✅ Improper values (e.g., negative population, GDP values of 0, etc.)

✅ Duplicate country-year entries

By doing this before saving the cleaned dataset, we avoid creating multiple intermediate files and ensure clean data from the start.

In [15]:
population.describe()

1960          1961          1962          1963          1964  \
count  2.640000e+02  2.640000e+02  2.640000e+02  2.640000e+02  2.640000e+02   
mean   1.154480e+08  1.170538e+08  1.192160e+08  1.218878e+08  1.245835e+08   
std    3.626523e+08  3.671660e+08  3.738303e+08  3.824608e+08  3.911397e+08   
min    2.715000e+03  2.970000e+03  3.264000e+03  3.584000e+03  3.922000e+03   
25%    5.152028e+05  5.255230e+05  5.363018e+05  5.475875e+05  5.593638e+05   
50%    3.659633e+06  3.747132e+06  3.831900e+06  3.919710e+06  4.010150e+06   
75%    2.686293e+07  2.761326e+07  2.837302e+07  2.915448e+07  2.995223e+07   
max    3.021529e+09  3.062769e+09  3.117373e+09  3.184063e+09  3.251253e+09   

               1965          1966          1967          1968          1969  \
count  2.640000e+02  2.640000e+02  2.640000e+02  2.640000e+02  2.640000e+02   
mean   1.273111e+08  1.301581e+08  1.330127e+08  1.359425e+08  1.389705e+08   
std    3.999256e+08  4.091870e+08  4.184361e+08  4.279507e+08  4.378239e+08   
min    4.282000e+03  4.664000e+03  5.071000e+03  5.500000e+03  5.631000e+03   
25%    5.675750e+05  5.711695e+05  5.779525e+05  5.825170e+05  5.861185e+05   
50%    4.102976e+06  4.198738e+06  4.297792e+06  4.396290e+06  4.503420e+06   
75%    3.075921e+07  3.147516e+07  3.203946e+07  3.247057e+07  3.277149e+07   
max    3.318998e+09  3.389087e+09  3.459014e+09  3.530702e+09  3.604812e+09   

       ...          2015          2016          2017          2018  \
count  ...  2.650000e+02  2.650000e+02  2.650000e+02  2.650000e+02   
mean   ...  3.004900e+08  3.043329e+08  3.081725e+08  3.119272e+08   
std    ...  9.388565e+08  9.499906e+08  9.611259e+08  9.719209e+08   
min    ...  1.095400e+04  1.093000e+04  1.086900e+04  1.075100e+04   
25%    ...  1.786457e+06  1.777568e+06  1.791019e+06  1.797086e+06   
50%    ...  1.035808e+07  1.032545e+07  1.025915e+07  1.028382e+07   
75%    ...  6.073058e+07  6.062750e+07  6.053671e+07  6.042176e+07   
max    ...  7.441472e+09  7.528523e+09  7.614114e+09  7.696495e+09   

               2019          2020          2021          2022          2023  \
count  2.650000e+02  2.650000e+02  2.650000e+02  2.650000e+02  2.650000e+02   
mean   3.156080e+08  3.192036e+08  3.223546e+08  3.255235e+08  3.287886e+08   
std    9.824034e+08  9.924599e+08  1.001248e+09  1.009781e+09  1.018589e+09   
min    1.058100e+04  1.039900e+04  1.019400e+04  9.992000e+03  9.816000e+03   
25%    1.788891e+06  1.790152e+06  1.786080e+06  1.803545e+06  1.827816e+06   
50%    1.042338e+07  1.069786e+07  1.050577e+07  1.048694e+07  1.064485e+07   
75%    5.972908e+07  6.097280e+07  6.283041e+07  6.471182e+07  6.661761e+07   
max    7.776892e+09  7.856139e+09  7.921184e+09  7.989982e+09  8.061876e+09   

       Unnamed: 68  
count          0.0  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN  

[8 rows x 65 columns]

In [16]:
gdp.describe()

1960          1961          1962          1963          1964  \
count  1.510000e+02  1.540000e+02  1.570000e+02  1.570000e+02  1.570000e+02   
mean   6.786171e+10  6.982251e+10  7.327492e+10  7.908218e+10  8.680328e+10   
std    2.019254e+11  2.116016e+11  2.257269e+11  2.427310e+11  2.657976e+11   
min    1.201202e+07  1.159202e+07  1.254164e+07  1.283330e+07  1.341663e+07   
25%    5.079241e+08  5.007338e+08  5.740911e+08  5.862949e+08  5.828164e+08   
50%    3.359404e+09  3.330233e+09  3.308913e+09  3.988462e+09  4.016794e+09   
75%    3.325071e+10  3.282977e+10  3.184162e+10  3.657288e+10  3.319881e+10   
max    1.371947e+12  1.445951e+12  1.550598e+12  1.669570e+12  1.830168e+12   

               1965          1966          1967          1968          1969  \
count  1.630000e+02  1.640000e+02  1.670000e+02  1.680000e+02  1.680000e+02   
mean   9.133482e+10  9.793328e+10  1.014971e+11  1.086794e+11  1.203480e+11   
std    2.844773e+11  3.088386e+11  3.257876e+11  3.515860e+11  3.876032e+11   
min    1.359393e+07  1.446908e+07  1.583511e+07  1.460000e+07  1.585000e+07   
25%    5.956572e+08  6.450667e+08  6.311234e+08  6.211906e+08  6.574670e+08   
50%    3.817227e+09  4.153527e+09  3.532700e+09  4.529031e+09  5.087251e+09   
75%    3.464910e+10  3.720117e+10  3.658090e+10  3.851340e+10  4.334334e+10   
max    1.994298e+12  2.161754e+12  2.293944e+12  2.478900e+12  2.738144e+12   

       ...          2015          2016          2017          2018  \
count  ...  2.590000e+02  2.580000e+02  2.580000e+02  2.580000e+02   
mean   ...  2.382043e+12  2.424134e+12  2.597268e+12  2.765285e+12   
std    ...  8.174322e+12  8.328694e+12  8.865044e+12  9.440413e+12   
min    ...  3.681194e+07  4.162906e+07  4.527660e+07  4.801526e+07   
25%    ...  8.766202e+09  8.620784e+09  9.193745e+09  9.928840e+09   
50%    ...  4.871750e+10  4.806565e+10  5.277001e+10  5.609719e+10   
75%    ...  4.973625e+11  5.042315e+11  5.341521e+11  5.491441e+11   
max    ...  7.547247e+13  7.670255e+13  8.171204e+13  8.688484e+13   

               2019          2020          2021          2022          2023  \
count  2.580000e+02  2.570000e+02  2.570000e+02  2.540000e+02  2.420000e+02   
mean   2.807353e+12  2.736278e+12  3.141029e+12  3.307885e+12  3.618640e+12   
std    9.580867e+12  9.368340e+12  1.069643e+13  1.114471e+13  1.190401e+13   
min    5.412320e+07  5.174659e+07  6.019641e+07  5.906598e+07  6.228031e+07   
25%    1.011166e+10  9.516738e+09  1.007135e+10  1.246361e+10  1.543170e+10   
50%    5.775248e+10  5.366864e+10  6.152928e+10  6.970415e+10  8.383752e+10   
75%    5.421641e+11  5.451476e+11  6.371869e+11  6.799031e+11  1.021922e+12   
max    8.814985e+13  8.576301e+13  9.784830e+13  1.017709e+14  1.061717e+14   

       Unnamed: 68  
count          0.0  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN  

[8 rows x 65 columns]

In [17]:
### 1️⃣ Check for Missing Values ###
print("\n📌 Checking Missing Values in Population Data:")
print(population.isnull().sum())

print("\n📌 Checking Missing Values in GDP Data:")
print(gdp.isnull().sum())


📌 Checking Missing Values in Population Data:
Country Name        0
Country Code        0
Indicator Name      0
Indicator Code      0
1960                2
                 ... 
2020                1
2021                1
2022                1
2023                1
Unnamed: 68       266
Length: 69, dtype: int64

📌 Checking Missing Values in GDP Data:
Country Name        0
Country Code        0
Indicator Name      0
Indicator Code      0
1960              115
                 ... 
2020                9
2021                9
2022               12
2023               24
Unnamed: 68       266
Length: 69, dtype: int64


Since there are some missing values in the emissions columns for some years, instead of removing missing values, we can use the forward-fill method (ffill()) to propagate the last valid value forward. This is a good approach because population and GDP typically follow a trend over time, meaning the previous year’s value is a reasonable estimate when data is missing. As for the negative values, we will replace the negative values with absolute values approach. 

In [18]:
# Drop unnecessary columns (keeping only 1990-2021)
cols_to_keep = ['Country Name', 'Country Code'] + [str(year) for year in range(1990, 2022)]
population = population[cols_to_keep]
gdp = gdp[cols_to_keep]

# Convert Wide Format to Long Format
population_long = population.melt(id_vars=['Country Name', 'Country Code'], var_name='Year', value_name='Population')
gdp_long = gdp.melt(id_vars=['Country Name', 'Country Code'], var_name='Year', value_name='GDP')

# Convert 'Year' to integer
population_long['Year'] = population_long['Year'].astype(int)
gdp_long['Year'] = gdp_long['Year'].astype(int)

### 3️⃣ Handle Negative & Missing Values (Same as Emissions Dataset) ###
# Replace negative values with NaN
population_long.loc[population_long['Population'] < 0, 'Population'] = None
gdp_long.loc[gdp_long['GDP'] < 0, 'GDP'] = None

# Forward-fill missing values
population_long['Population'].fillna(method='ffill', inplace=True)
gdp_long['GDP'].fillna(method='ffill', inplace=True)

# Check if negatives still exist
print("\n📌 Min Population Value After Cleaning:", population_long['Population'].min())
print("\n📌 Min GDP Value After Cleaning:", gdp_long['GDP'].min())

### Check for Duplicate Entries ###
print("\n📌 Checking for Duplicate Country-Year Entries in Population Data:")
print(population_long.duplicated(subset=['Country Code', 'Year']).sum())

print("\n📌 Checking for Duplicate Country-Year Entries in GDP Data:")
print(gdp_long.duplicated(subset=['Country Code', 'Year']).sum())

# Drop duplicate records (if any)
population_long.drop_duplicates(subset=['Country Code', 'Year'], inplace=True)
gdp_long.drop_duplicates(subset=['Country Code', 'Year'], inplace=True)

### 5️⃣ Rename Columns for Consistency ###
population_long.rename(columns={'Country Name': 'Country', 'Country Code': 'ISO'}, inplace=True)
gdp_long.rename(columns={'Country Name': 'Country', 'Country Code': 'ISO'}, inplace=True)

### 6️⃣ Save the Cleaned & Filtered Data ###
population_long.to_csv("filtered_population.csv", index=False)
gdp_long.to_csv("filtered_gdp.csv", index=False)

print("\n✅ Data Cleaning Completed. Filtered Population & GDP Data Saved for 1990-2021.")


📌 Min Population Value After Cleaning: 8798.0

📌 Min GDP Value After Cleaning: 9542900.90136505

📌 Checking for Duplicate Country-Year Entries in Population Data:
0

📌 Checking for Duplicate Country-Year Entries in GDP Data:
0

✅ Data Cleaning Completed. Filtered Population & GDP Data Saved for 1990-2021.


/var/folders/rg/j1hzfkkx1fn412ygjvnwy86h0000gn/T/ipykernel_77281/129892087.py:20: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  population_long['Population'].fillna(method='ffill', inplace=True)
/var/folders/rg/j1hzfkkx1fn412ygjvnwy86h0000gn/T/ipykernel_77281/129892087.py:21: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  gdp_long['GDP'].fillna(method='ffill', inplace=True)
